dpkg and dpkg_src rules for deb packages

r2d4 opened the pull request:
These are bazel rules to help resolve remote locations of the .deb
dependencies in our images.

ref #7 

This is a first pass at adding rules that make the .deb resolution more user friendly and less fragile.  No need to resolve the URLs manually and hardcode into an http_rule.  I'm not sure how to really incorporate them in the WORKSPACE, but I imagined something like this in BUILD files.

```
load("//packages:deb_pkg.bzl", "deb_pkg")
load("//packages:deb_src.bzl", "deb_src")

deb_pkg(
    name = "openssl",
    pkg_name = "openssl",
    source = ":debian_jessie",
)

deb_src(
    name = "debian_jessie",
    arch = "amd64",
    distro = "jessie",
    mirror_url = "http://httpredir.debian.org",
)
```

then you can reference the debs in your docker_rule by 

```
docker_build(
    name = "base",
    base = ":with_tmp",
    debs = [
        ":openssl",
        ...
    ],
)
```

This could also the be the entrypoint for future deb rules like dependency resolution.  I probably need to add version checking and checksum verification to these rules also.

These are my first bazel rules so let me know if I'm on the right track :)

mattmoor commented on packages/deb_pkg.bzl:
```diff
@@ -0,0 +1,31 @@
+    outputs = [ctx.outputs.deb],
+  )
+
+deb_pkg = rule(
```
We want this to be a `repository_rule`, so that it can run in `WORKSPACE`, which brings with it new challenges (e.g. you can't depend on building the tool you plan to run).

Let's talk about this more once you are back, but what I do for `docker_pull` is publish a PAR file.
 - cloudbuild to produce it on release tags: https://github.com/google/containerregistry/blob/master/puller.yaml
 - where the binary is pulled down: https://github.com/bazelbuild/rules_docker/blob/master/docker/docker.bzl#L32
 - where the binary is used: https://github.com/bazelbuild/rules_docker/blob/master/docker/pull.bzl#L67

Once this is working, we should consider where the best place for this rule would be.  @damienmg FYI

mmikulicic commented:
cool; I was about doing something similar then I found this

shouldn't we force a version and/or sha of the `deb_pkg` being referenced (in order to have correct incremental and reproducible builds)?

mattmoor commented:
@mmikulicic Yes, for reproducible builds, you'd want a checksum.  I generally call this "pseudo-hermetic" because while it's not hermetic you always get the same thing (or fail due to availability).  Sort of like `-Werror` for reproducibility I guess.

mmikulicic commented:
@mattmoor I don't know the exact definition of hermeticity that everybody can agree on, but I thought the interesting property is that tooling can reason about producing outputs as a pure function of the inputs by having a complete visibility of all the components that are touched while doing a build (including the tooling itself).

If something is a local file or a remote file surely has practical implications on how likely is to fail, but from a theoretical point of view a local file is no different: the disk I/O can file as well :-)

Or did I misunderstand your comment?

mattmoor commented:
@mmikulicic So hermetic is an effectively closed environment, e.g. restricted to your mono-repo.  This would be Bazel without a `WORKSPACE` at all, as we use it internally.

When references reach outside of your not-so-mono-repo, they can be either mutable (in the sense of different versions over time) or immutable (maybe it can go away, but not change).

So when I say "pseudo-hermetic", what I mean is that everything you need is contained within your repo EXCEPT stuff (transitively) referenced in an immutable way.  

So it's sort of hermetic in the sense that your repo *defines* the world exactly (reproducibly), but doesn't fully *contain* it.

mmikulicic commented:
I always wondered why bazel's `local_repository` requires an absolute path; with a relative path it would be straightforward to vendor other repositories thus making it easier to achieve this kind of fully local hermeticity.

mattmoor commented:
@mmikulicic I think the docs say absolute, but in at least one failed (for other reasons) experiment, I was able to use relative paths w/ `local_repository`.

Consider opening an issue on `bazelbuild/bazel` to discuss?  Feel free to CC me.

r2d4 commented:
Sorry for the delay - I've updated this PR with a few things:

* Split into two commits for now - one introducing the rules and one using the rules in the distroless images.  This is because these rules might not live in this repo.
* The dpkg parser is now written in python rather than go.  Since we are building and uploading binaries, using go would have required uploading 3 binaries and doing platform detection which turned out to be a little tricky.

Future TODOs:

* Support specifying a version in the dpkg rule
* Support expanding dependencies for packages (This may be trickier than it seems)
* Checksum validation for dpkg_src rule.  We can validate checksums from the Packages.gz file in the dpkg rule.
* Not reusing the rule name for the package name, since not all package names are valid rule names (e.g. libstdc++)

cc @dlorenc @mattmoor

dlorenc commented on package_manager/cloudbuild.yaml:
```diff
@@ -0,0 +1,18 @@
+steps:
+- name: gcr.io/cloud-builders/bazel
```
Could you add a few file-level comments explaining what this is for?

r2d4 replied:
done

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,99 @@
+import argparse
```
I think we need a copyright string up here, also a file-level comment explaining this.

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,99 @@
+        metadata = json.load(f)
+    for pkg in metadata:
+        if pkg["Package"] == package_name:
+            print(pkg)
```
Probably drop the print?

r2d4 replied:
fixed

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,99 @@
+def download_dpkg(packages_file, package_name):
+    with open(packages_file, 'rb') as f:
+        metadata = json.load(f)
+    for pkg in metadata:
```
Maybe add an else: to the for loop to fail if we never find the package?

r2d4 replied:
fixed

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,99 @@
+        distro,
+        arch
+    )
+    print(url)
```
Drop the print?

r2d4 replied:
fixed

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,99 @@
+            break
+
+def download_package_list(mirror_url, distro, arch):
+    url = "%s/debian/dists/%s/main/binary-%s/Packages.gz" % (
```
path.join is probably a bit safer.

r2d4 replied:
will path.join work on windows?  Not too familiar with the python impl but can it also be used for URLs?

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,99 @@
+    with open(PACKAGES_FILE_NAME, 'w') as f:
+        json.dump(metadata, f)
+ 
+def parse_package_metadata(data, separator):
```
is the separator ever not a colon?

r2d4 replied:
Some other package managers other than debian/apt-get use a different separator. I added it as a constant instead.

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,99 @@
+    data = f.read()
+    metadata = parse_package_metadata(data, ":")
+    for pkg in metadata:
+        pkg["Filename"] = mirror_url + "/debian/" + pkg["Filename"]
```
path.join

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,99 @@
+ 
+def parse_package_metadata(data, separator):
+    metadata = []
+    found = False
```
found looks unused.

r2d4 replied:
fixed

mattmoor commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,113 @@
+    with open(PACKAGES_FILE_NAME, 'w') as f:
+        json.dump(metadata, f)
+ 
+def parse_package_metadata(data):
```
We probably want this in a py_library we can write unit tests for.

r2d4 commented:
I split `parse_metadata` into its own py_lib and added a few unit tests.  Pending more review feedback, the only thing left to do is it figure out where we can upload `dpkg_parser.par`, maybe into a distroless bucket for `distroless` @mattmoor ?

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,75 @@
+    f = gzip.GzipFile(fileobj=io.BytesIO(buf.read()))
+    data = f.read()
+    metadata = parse_package_metadata(data, mirror_url)
+    with open(PACKAGES_FILE_NAME, 'w') as f:
```
Some comments explaining the schema of this JSON file and explaining why we need it at all would be useful.

r2d4 replied:
I added a few more comments throughout, and docstrings to each function

dlorenc commented on package_manager/parse_metadata_test.py:
```diff
@@ -0,0 +1,33 @@
+                             "newmail",
+                             "zzuf",]
+        for pkg in expected_packages:
+            ok = self.metadata[pkg]
```
What is this supposed to do, just check that the package is in metadata? You can use something like self.assertContains instead.

r2d4 replied:
Yep, switched it over to self.assertContains

dlorenc commented on package_manager/cloudbuild.yaml:
```diff
@@ -0,0 +1,30 @@
+# A cloud build config to release a PAR binary of
+# the dpkg_parser.par used for the package manager bazel rules
+
+# Takes a substitution of _VERSION
```
Does this substitution work via triggers? We can set this up to run automatically everytime a PR is merged.

r2d4 replied:
I added a trigger on the project for the master branch and versioned it with $COMMIT_SHA.  Not sure if this is the best solution, since we really don't care if its built on _every_ commit, but tags are probably not appropriate since this is just a tool that lives in the distroless repo.

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,109 @@
+
+
+def download_dpkg(packages_file, package_name):
+    """ Using an unzipped, json package file with full urls, downloads a .deb package 
```
Nit: no spaces before or after """. We might want to enable pylint via travis.

r2d4 replied:
Fixed

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -0,0 +1,109 @@
+    
+    Uses the 'Filename' key to download the .deb package
+    """
+    found = False
```
found looks unused.

r2d4 replied:
Fixed

dlorenc commented on package_manager/parse_metadata_test.py:
```diff
@@ -0,0 +1,33 @@
+class TestParseMetadata(unittest.TestCase):
+
+    def setUp(self):
+        dir = os.path.dirname(__file__)
```
nit: dir is technically a builtin you're shadowing here. A different name would be better.

r2d4 replied:
Fixed
